packages/modules/promotion/src/utils/compute-actions/buy-get.ts

```
@@ -260,21 +320,244 @@ export function getComputedActionsForBuyGet(
+  methodIdPromoValueMap: Map<string, BigNumberInput>,
+  eligibleBuyItemMap: Map<string, EligibleItem[]>,
+  eligibleTargetItemMap: Map<string, EligibleItem[]>
+): PromotionTypes.ComputeActions[] {
+  if (!isValidPromotionContext(promotion, itemsContext)) {
+    return []
+  }
+
+  const applicationConfig =
+    normalizePromotionApplicationConfiguration(promotion)
+
+  const itemsMap = new Map<string, ComputeActionItemLine>(
+    itemsContext.map((i) => [i.id, i])
+  )
+
+  const eligibleBuyItems = filterItemsByPromotionRules(
+    itemsContext,
+    promotion.application_method?.buy_rules
+  )
+
+  const eligibleTargetItems = filterItemsByPromotionRules(
+    itemsContext,
+    promotion.application_method?.target_rules
+  )
+
+  const remainingBuyQuantities = calculateRemainingQuantities(
+    eligibleBuyItems,
+    eligibleBuyItemMap,
+    promotion.code!
+  )
+
+  const remainingTargetQuantities = calculateRemainingQuantities(
+    eligibleTargetItems,
+    eligibleTargetItemMap,
+    promotion.code!
+  )
+
+  const totalEligibleBuyItemsMap = new Map<string, EligibleItem>()
+  const totalEligibleTargetItemsMap = new Map<string, EligibleItem>()
+  const itemIdPromotionAmountMap = new Map<string, BigNumberInput>()
+  const computedActions: PromotionTypes.ComputeActions[] = []
+
+  let appliedPromotionQuantity = MathBN.convert(0)
+
+  /*
+    This loop continues applying the promotion until one of the stopping conditions is met:
+    - No more items satisfy the minimum buy quantity requirement
+    - Maximum applicable promotion quantity is reached  
+    - No valid target items can be found for promotion application
+    
+    Each iteration:
+    1. Prepares an application state (selects buy items + eligible target items)
+    2. Applies promotion to the selected target items
+    3. Updates remaining quantities to prevent double-usage in next iteration
+    4. Updates the total eligible items for next iteration
+  */
+  while (true) {
```

nit: I saw other while (true) in the codebase so feel free to disregard this comment, but in my experience, bad things eventually happen in while (true), even when we think that we cover all the use cases for the loop the end. I suggest using a max counter (something ridiculously high for this flow) and if we reach the max counter, we can throw an error

---

good call, added it in there